### ISSUE TITLE
Add unit tests for eviction logic

### ISSUE THREAD

**Issue by practicalswift** [bitcoin/bitcoin#19966]
The eviction logic is one of the most important parts in our code base: a bug in the eviction logic could allow for attacker triggered network partitioning.

Unfortunately the eviction logic is currently not tested properly.

The unit tests and fuzz tests don't cover the eviction logic at all.

The functional tests cover some parts of the eviction logic but the coverage is spotty/incomplete.

Unfortunately `CConnman::AttemptToEvictConnection` is currently somewhat tricky to test.

To solve this I suggest simplifying `CConnman::AttemptToEvictConnection()` by introducing a new function ...

```
[[nodiscard]] Optional<NodeEvictionCandidate> SelectNodeToEvict(std::vector<NodeEvictionCandidate> vEvictionCandidates)
```

... which would be called from the now much simplified `CConnman::AttemptToEvictConnection`.

`SelectNodeToEvict` would contain all eviction selection logic and could thus be tested trivially.

**Comment by jonatack** [comment]
Concept ACK on improving the design testability and coverage here.

**Comment by practicalswift** [comment]
> The functional tests cover some parts of the eviction logic but the coverage is spotty/incomplete.

In order to get a rough sense for the quality of the current eviction logic testing I performed some basic mutation testing of this code.

All tests (including `p2p_eviction.py`) pass successfully also in the presence of these mutations:

```diff
$ git diff
diff --git a/src/net.cpp b/src/net.cpp
index e35d05cec..e138a4a13 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -826,43 +826,36 @@ static bool ReverseCompareNodeMinPingTime(const NodeEvictionCandidate &a, const

 static bool ReverseCompareNodeTimeConnected(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
-    return a.nTimeConnected > b.nTimeConnected;
+    return false;
 }

 static bool CompareLocalHostTimeConnected(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
-    if (a.m_is_local != b.m_is_local) return b.m_is_local;
-    return a.nTimeConnected > b.nTimeConnected;
+    return false;
 }

 static bool CompareNetGroupKeyed(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b) {
-    return a.nKeyedNetGroup < b.nKeyedNetGroup;
+    return false;
 }

 static bool CompareNodeBlockTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
     // There is a fall-through here because it is common for a node to have many peers which have not yet relayed a block.
     if (a.nLastBlockTime != b.nLastBlockTime) return a.nLastBlockTime < b.nLastBlockTime;
-    if (a.fRelevantServices != b.fRelevantServices) return b.fRelevantServices;
-    return a.nTimeConnected > b.nTimeConnected;
+    return false;
 }

 static bool CompareNodeTXTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
     // There is a fall-through here because it is common for a node to have more than a few peers that have not yet relayed txn.
     if (a.nLastTXTime != b.nLastTXTime) return a.nLastTXTime < b.nLastTXTime;
-    if (a.fRelayTxes != b.fRelayTxes) return b.fRelayTxes;
-    if (a.fBloomFilter != b.fBloomFilter) return a.fBloomFilter;
-    return a.nTimeConnected > b.nTimeConnected;
+    return false;
 }

 // Pick out the potential block-relay only peers, and sort them by last block time.
 static bool CompareNodeBlockRelayOnlyTime(const NodeEvictionCandidate &a, const NodeEvictionCandidate &b)
 {
-    if (a.fRelayTxes != b.fRelayTxes) return a.fRelayTxes;
-    if (a.nLastBlockTime != b.nLastBlockTime) return a.nLastBlockTime < b.nLastBlockTime;
-    if (a.fRelevantServices != b.fRelevantServices) return b.fRelevantServices;
-    return a.nTimeConnected > b.nTimeConnected;
+    return false;
 }

 //! Sort an array by the specified comparator, then erase the last K elements.
@@ -870,8 +863,6 @@ template<typename T, typename Comparator>
 static void EraseLastKElements(std::vector<T> &elements, Comparator comparator, size_t k)
 {
     std::sort(elements.begin(), elements.end(), comparator);
-    size_t eraseSize = std::min(k, elements.size());
-    elements.erase(elements.end() - eraseSize, elements.end());
 }

 /** Try to find a connection to evict when the node is full.
$ make -C src/ bitcoind
$ make check
$ echo $?
0
$ test/functional/test_runner.py
$ echo $?
0
```

**Comment by practicalswift** [comment]
The small patch taking us to the suggested unit testable `SelectNodeToEvict` could look something like this:

```patch
diff --git a/src/net.cpp b/src/net.cpp
index 79e7f88c2..10c1246e4 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -874,43 +874,7 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
     elements.erase(elements.end() - eraseSize, elements.end());
 }
 
-/** Try to find a connection to evict when the node is full.
- *  Extreme care must be taken to avoid opening the node to attacker
- *   triggered network partitioning.
- *  The strategy used here is to protect a small number of peers
- *   for each of several distinct characteristics which are difficult
- *   to forge.  In order to partition a node the attacker must be
- *   simultaneously better at all of them than honest peers.
- */
-bool CConnman::AttemptToEvictConnection()
-{
-    std::vector<NodeEvictionCandidate> vEvictionCandidates;
-    {
-        LOCK(cs_vNodes);
-
-        for (const CNode* node : vNodes) {
-            if (node->HasPermission(PF_NOBAN))
-                continue;
-            if (!node->IsInboundConn())
-                continue;
-            if (node->fDisconnect)
-                continue;
-            bool peer_relay_txes = false;
-            bool peer_filter_not_null = false;
-            if (node->m_tx_relay != nullptr) {
-                LOCK(node->m_tx_relay->cs_filter);
-                peer_relay_txes = node->m_tx_relay->fRelayTxes;
-                peer_filter_not_null = node->m_tx_relay->pfilter != nullptr;
-            }
-            NodeEvictionCandidate candidate = {node->GetId(), node->nTimeConnected, node->nMinPingUsecTime,
-                                               node->nLastBlockTime, node->nLastTXTime,
-                                               HasAllDesirableServiceFlags(node->nServices),
-                                               peer_relay_txes, peer_filter_not_null, node->addr, node->nKeyedNetGroup,
-                                               node->m_prefer_evict, node->addr.IsLocal()};
-            vEvictionCandidates.push_back(candidate);
-        }
-    }
-
+[[nodiscard]] Optional<NodeEvictionCandidate> SelectNodeToEvict(std::vector<NodeEvictionCandidate> vEvictionCandidates) {
     // Protect connections with certain characteristics
 
     // Deterministically select 4 peers to protect by netgroup.
@@ -948,7 +912,7 @@ bool CConnman::AttemptToEvictConnection()
     total_protect_size -= initial_size - vEvictionCandidates.size();
     EraseLastKElements(vEvictionCandidates, ReverseCompareNodeTimeConnected, total_protect_size);
 
-    if (vEvictionCandidates.empty()) return false;
+    if (vEvictionCandidates.empty()) return nullopt;
 
     // If any remaining peers are preferred for eviction consider only them.
     // This happens after the other preferences since if a peer is really the best by other criteria (esp relaying blocks)
@@ -979,8 +943,56 @@ bool CConnman::AttemptToEvictConnection()
     // Reduce to the network group with the most connections
     vEvictionCandidates = std::move(mapNetGroupNodes[naMostConnections]);
 
+    if (vEvictionCandidates.empty()) {
+        return nullopt;
+    }
+
     // Disconnect from the network group with the most connections
-    NodeId evicted = vEvictionCandidates.front().id;
+    return vEvictionCandidates.front();
+}
+
+/** Try to find a connection to evict when the node is full.
+ *  Extreme care must be taken to avoid opening the node to attacker
+ *   triggered network partitioning.
+ *  The strategy used here is to protect a small number of peers
+ *   for each of several distinct characteristics which are difficult
+ *   to forge.  In order to partition a node the attacker must be
+ *   simultaneously better at all of them than honest peers.
+ */
+bool CConnman::AttemptToEvictConnection()
+{
+    std::vector<NodeEvictionCandidate> vEvictionCandidates;
+    {
+        LOCK(cs_vNodes);
+
+        for (const CNode* node : vNodes) {
+            if (node->HasPermission(PF_NOBAN))
+                continue;
+            if (!node->IsInboundConn())
+                continue;
+            if (node->fDisconnect)
+                continue;
+            bool peer_relay_txes = false;
+            bool peer_filter_not_null = false;
+            if (node->m_tx_relay != nullptr) {
+                LOCK(node->m_tx_relay->cs_filter);
+                peer_relay_txes = node->m_tx_relay->fRelayTxes;
+                peer_filter_not_null = node->m_tx_relay->pfilter != nullptr;
+            }
+            NodeEvictionCandidate candidate = {node->GetId(), node->nTimeConnected, node->nMinPingUsecTime,
+                                               node->nLastBlockTime, node->nLastTXTime,
+                                               HasAllDesirableServiceFlags(node->nServices),
+                                               peer_relay_txes, peer_filter_not_null, node->addr, node->nKeyedNetGroup,
+                                               node->m_prefer_evict, node->addr.IsLocal()};
+            vEvictionCandidates.push_back(candidate);
+        }
+    }
+
+    const Optional<NodeEvictionCandidate> node_to_evict = SelectNodeToEvict(vEvictionCandidates);
+    if (!node_to_evict) {
+        return false;
+    }
+    const NodeId evicted = node_to_evict->id;
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
         if (pnode->GetId() == evicted) {
```

**Comment by amitiuttarwar** [comment]
strong concept ACK on improving eviction testing 🙌🏽 . its critical logic, hard to reason about, and lacking automated testing. 

I'm trying to wrap my head around this strategy, is this right so far? - 
- the main idea is to separate `AttemptToEvictConnection` into two parts of 1. selecting eviction candidates 2. selecting which node to evict from the candidates
- the proposed strategy is to extract # 2 into `SelectNodeToEvict`.
- this enables sanity testing `SelectNodeToEvict` (as done in #19972) by constructing `NodeEvictionCandidates` with fuzzed inputs

this seems like a great start. do you think this would enable us to write unit tests as well?

**Comment by practicalswift** [comment]
> I'm trying to wrap my head around this strategy, is this right so far? -
> 
> * the main idea is to separate `AttemptToEvictConnection` into two parts of 1. selecting eviction candidates 2. selecting which node to evict from the candidates
> * the proposed strategy is to extract # 2 into `SelectNodeToEvict`.
> * this enables sanity testing `SelectNodeToEvict` (as done in #19972) by constructing `NodeEvictionCandidates` with fuzzed inputs

Exactly! :)

And as you point out the fuzzing harness is only meant as basic sanity testing to make sure the basic hygiene routine is covered -- that no ASan or UBSan warnings are triggered no matter what input is feed, and some very basic sanity assertions. The important testing of the eviction logic is expected to be done in unit tests.

> this seems like a great start. do you think this would enable us to write unit tests as well?

Absolutely! :)

I suggest introducing something along the lines of:

```c++
void AssertIsEvicted(const NodeId expected_to_be_evicted, std::vector<NodeEvictionCandidate> eviction_candidates);
void AssertIsNotEvicted(const NodeId expected_not_to_be_evicted, std::vector<NodeEvictionCandidate> eviction_candidates);
```

These could be used to test that the "correct nodes" are evicted/not evicted (whichever makes more sense) given some synthetic `NodeEvictionCandidate`:s carefully constructed to have characteristics that are relevant from an eviction policy perspective.

**Comment by ariard** [comment]
Concept ACK. I think covering heuristic stubs correctness (e.g `CompareLocalHostTimeConnected`) is a good step forward. These stubs aren't straightforward to understand so unit test coverage would be great.

I can't think of issue as splitting `AttemptToEvictConnection` between `SelectNodeToEvict`/`AttemptToEvictCandidates`.

**Comment by practicalswift** [comment]
This issue is addressed in #20477 ("test/net: Add unit testing of node eviction logic").
